File: solana/tests/helpers/tokenBridge/legacy/instructions/completeTransfer/native.ts
import { ParsedVaa } from "@certusone/wormhole-sdk";
import { BN } from "@coral-xyz/anchor";
import { TOKEN_PROGRAM_ID, getAssociatedTokenAddressSync } from "@solana/spl-token";
import {
  AccountMeta,
  PublicKey,
  SYSVAR_RENT_PUBKEY,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { TokenBridgeProgram, coreBridgeProgramId } from "../../..";
import { Claim, PostedVaaV1 } from "../../../../coreBridge";
import { Config, RegisteredEmitter, custodyAuthorityPda, custodyTokenPda } from "../../state";

export type LegacyCompleteTransferNativeContext = {
  payer: PublicKey;
  config?: PublicKey; // TODO: demonstrate this isn't needed in tests
  postedVaa?: PublicKey;
  claim?: PublicKey;
  registeredEmitter?: PublicKey;
  recipientToken: PublicKey;
  payerToken?: PublicKey;
  custodyToken?: PublicKey;
  mint: PublicKey;
  custodyAuthority?: PublicKey;
  recipient?: PublicKey | null;
  coreBridgeProgram?: PublicKey;
};

export function legacyCompleteTransferNativeAccounts(
  program: TokenBridgeProgram,
  accounts: LegacyCompleteTransferNativeContext,
  parsedVaa: ParsedVaa,
  overrides: {
    legacyRegisteredEmitterDerive: boolean;
  }
): LegacyCompleteTransferNativeContext {
  const programId = program.programId;
  const { emitterChain, emitterAddress, sequence, hash } = parsedVaa;

  const { legacyRegisteredEmitterDerive } = overrides;

  let {
    payer,
    config,
    postedVaa,
    claim,
    registeredEmitter,
    recipientToken,
    payerToken,
    custodyToken,
    mint,
    custodyAuthority,
    recipient,
    coreBridgeProgram,
  } = accounts;

  if (coreBridgeProgram === undefined) {
    coreBridgeProgram = coreBridgeProgramId(program);
  }

  if (config === undefined) {
    config = Config.address(programId);
  }

  if (postedVaa === undefined) {
    postedVaa = PostedVaaV1.address(coreBridgeProgram, Array.from(hash));
  }

  if (claim === undefined) {
    claim = Claim.address(
      programId,
      Array.from(emitterAddress),
      emitterChain,
      new BN(sequence.toString())
    );
  }

  if (registeredEmitter === undefined) {
    registeredEmitter = RegisteredEmitter.address(
      programId,
      emitterChain,
      legacyRegisteredEmitterDerive ? Array.from(emitterAddress) : undefined
    );
  }

  if (payerToken === undefined) {
    payerToken = getAssociatedTokenAddressSync(mint, payer);
  }

  if (custodyToken === undefined) {
    custodyToken = custodyTokenPda(programId, mint);
  }

  if (custodyAuthority === undefined) {
    custodyAuthority = custodyAuthorityPda(programId);
  }

  if (recipient === undefined) {
    recipient = SYSVAR_RENT_PUBKEY;
  } else if (recipient === null) {
    recipient = programId;
  }

  return {
    payer,
    config,
    postedVaa,
    claim,
    registeredEmitter,
    recipientToken,
    payerToken,
    custodyToken,
    mint,
    custodyAuthority,
    recipient,
    coreBridgeProgram,
  };
}

export function legacyCompleteTransferNativeIx(
  program: TokenBridgeProgram,
  accounts: LegacyCompleteTransferNativeContext,
  parsedVaa: ParsedVaa,
  overrides: {
    legacyRegisteredEmitterDerive?: boolean;
  } = {}
) {
  let { legacyRegisteredEmitterDerive } = overrides;

  if (legacyRegisteredEmitterDerive === undefined) {
    legacyRegisteredEmitterDerive = true;
  }

  const {
    payer,
    config,
    postedVaa,
    claim,
    registeredEmitter,
    recipientToken,
    payerToken,
    custodyToken,
    mint,
    custodyAuthority,
    recipient,
    coreBridgeProgram,
  } = legacyCompleteTransferNativeAccounts(program, accounts, parsedVaa, {
    legacyRegisteredEmitterDerive,
  });

  const keys: AccountMeta[] = [
    {
      pubkey: payer,
      isWritable: true,
      isSigner: true,
    },
    {
      pubkey: config,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: postedVaa,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: claim,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: registeredEmitter,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: recipientToken,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: payerToken,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: custodyToken,
      isWritable: true,
      isSigner: false,
    },
    {
      pubkey: mint,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: custodyAuthority,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: recipient,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: SystemProgram.programId,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: coreBridgeProgram,
      isWritable: false,
      isSigner: false,
    },
    {
      pubkey: TOKEN_PROGRAM_ID,
      isWritable: false,
      isSigner: false,
    },
  ];

  const data = Buffer.alloc(1, 2);

  return new TransactionInstruction({
    keys,
    programId: program.programId,
    data,
  });
}

File: solana/programs/token-bridge/src/legacy/instruction/complete_transfer/native.rs
#[cfg(feature = "no-entrypoint")]
mod __no_entrypoint {
    use crate::legacy::instruction::LegacyInstruction;
    use solana_program::{
        instruction::{AccountMeta, Instruction},
        pubkey::Pubkey,
    };

    pub struct CompleteTransferNative {
        pub payer: Pubkey,
        pub posted_vaa: Pubkey,
        pub claim: Pubkey,
        pub registered_emitter: Pubkey,
        pub recipient_token: Pubkey,
        pub payer_token: Pubkey,
        pub custody_token: Pubkey,
        pub mint: Pubkey,
        pub custody_authority: Pubkey,
        pub recipient: Option<Pubkey>,
        pub system_program: Pubkey,
        pub core_bridge_program: Pubkey,
        pub token_program: Pubkey,
    }

    pub fn complete_transfer_native(accounts: CompleteTransferNative) -> Instruction {
        let recipient = accounts.recipient.unwrap_or(crate::ID);

        let accounts = vec![
            AccountMeta::new(accounts.payer, true),
            AccountMeta::new_readonly(crate::ID, false), // _config
            AccountMeta::new_readonly(accounts.posted_vaa, false),
            AccountMeta::new(accounts.claim, false),
            AccountMeta::new_readonly(accounts.registered_emitter, false),
            AccountMeta::new(accounts.recipient_token, false),
            AccountMeta::new(accounts.payer_token, false),
            AccountMeta::new(accounts.custody_token, false),
            AccountMeta::new_readonly(accounts.mint, false),
            AccountMeta::new_readonly(accounts.custody_authority, false),
            AccountMeta::new_readonly(recipient, false),
            AccountMeta::new_readonly(accounts.system_program, false),
            AccountMeta::new_readonly(accounts.core_bridge_program, false),
            AccountMeta::new_readonly(accounts.token_program, false),
        ];

        Instruction::new_with_borsh(
            crate::ID,
            &(LegacyInstruction::CompleteTransferNative),
            accounts,
        )
    }
}

#[cfg(feature = "no-entrypoint")]
pub use __no_entrypoint::*;
